packages/smooth_app/lib/pages/personalized_ranking_page.dart

```diff
@@ -37,52 +39,103 @@ class _PersonalizedRankingPageState extends State<PersonalizedRankingPage>
+  }
+
   @override
   Widget build(BuildContext context) {
     final ProductPreferences productPreferences =
         context.watch<ProductPreferences>();
-    final LocalDatabase localDatabase = context.watch<LocalDatabase>();
-    final DaoProductList daoProductList = DaoProductList(localDatabase);
-    final List<MatchedProductV2> allProducts = MatchedProductV2.sort(
-      widget.products,
-      productPreferences,
-    );
-    final AppLocalizations appLocalizations = AppLocalizations.of(context);
-
-    AnalyticsHelper.trackPersonalizedRanking(widget.products.length);
-
-    MatchedProductStatusV2? status;
-    final List<_VirtualItem> list = <_VirtualItem>[];
-    for (final MatchedProductV2 product in allProducts) {
-      if (status == null || status != product.status) {
-        status = product.status;
-        list.add(_VirtualItem.status(status));
-      }
-      list.add(_VirtualItem.product(product));
-    }
-    final bool darkMode = Theme.of(context).brightness == Brightness.dark;
-    return SmoothScaffold(
-      appBar: AppBar(
-        title: Text(
-          widget.title,
-          overflow: TextOverflow.fade,
+    return Consumer<UpToDateProductProvider>(
+      builder: (
+        final BuildContext context,
+        final UpToDateProductProvider upToDateProductProvider,
+        final Widget? child,
+      ) =>
+          SmoothScaffold(
+        appBar: AppBar(
+          title: Text(widget.title, overflow: TextOverflow.fade),
         ),
-      ),
-      body: ListView.builder(
-        itemCount: list.length,
-        itemBuilder: (BuildContext context, int index) => _buildItem(
-          list[index],
-          daoProductList,
-          appLocalizations,
-          darkMode,
+        body: ChangeNotifierProvider<PersonalizedRankingModel>(
+          create: (final BuildContext context) => _model,
+          builder: (final BuildContext context, final Widget? wtf) {
+            context.watch<PersonalizedRankingModel>();
+            final List<String> compactPreferences =
+                productPreferences.getCompactView();
+            if (_compactPreferences == null) {
+              _compactPreferences = compactPreferences;
+              _model.refresh(context.read<LocalDatabase>(), productPreferences);
+            } else {
+              bool refresh = !_compactPreferences!.equals(compactPreferences);
+              if (!refresh) {
+                refresh = _model.needsRefresh(upToDateProductProvider);
+              }
+              if (refresh) {
+                return Center(
+                  child: Padding(
+                    padding: const EdgeInsets.all(SMALL_SPACE),
+                    child: ElevatedButton.icon(
+                      icon: const Icon(Icons.refresh),
+                      // TODO(monsieurtanuki): localize
+                      label: const Text(
+                          'Refresh the list with the new preferences'),
```

Comment:
Lets get in the habbit of directly localizing so that hopefully there is already a translation for the I think currently weekly release

Reply:
I don' agree with you on that.
While I'm developing, I don't know what will be kept in the end. And I don't even have time to check if other existing translations may match my requirement.
So instead of loading the `app_en.arb` file with temporary crap items with not really appropriate ids because I have no time to waste about it _at developing time_, I prefer to put a `TODO`.
If the code is kept, then I can think of the translation item.
If the code is not kept, it's removed and so is the `TODO`.

That said, as it looks like this piece of code is kept, now I need to put it in `app_en.arb` :)

Reply:
> I don't agree with you on that. While I'm developing, I don't know what will be kept in the end. And I don't even have time to check if other existing translations may match my requirement. So instead of loading the `app_en.arb` file with temporary crap items with not really appropriate ids because I have no time to waste about it _at developing time_, I prefer to put a `TODO`. If the code is kept, then I can think of the translation item. If the code is not kept, it's removed and so is the `TODO`.
> 
> That said, as it looks like this piece of code is kept, now I need to put it in `app_en.arb` :)
> 
> [Edit: added] And by the way there's a fair chance that this button is going to be removed in a very near future, for UX reasons.